pkg/apis/pipeline/v1beta1/pipelineparams_types.go

```diff
@@ -0,0 +1,131 @@
+/*
+Copyright 2018 The Knative Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1beta1
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// PipelineParamsSpec is the spec for a Pipeline resource
+type PipelineParamsSpec struct {
+	ServiceAccount string          `json:"serviceAccount"`
+	Sources        []Source        `json:"sources"`
+	ArtifactStores []ArtifactStore `json:"artifactStores"`
+	Results        Results         `json:"results"`
+}
+
+// Source is an endpoint from which to get data which is required
+// by a Build/Task for context (e.g. a repo from which to build an image).
+type Source struct {
+	Name           string `json:"name"`
+	Type           string `json:"type"`
+	URL            string `json:"url"`
+	Branch         string `json:"branch"`
+	Commit         string `json:"commit,omitempty"`
+	ServiceAccount string `json:"serviceAccount,omitempty"`
+}
```

Comment:
is this only intended to support git commits or gcs buckets? maybe a generic interface with git and gcs types implementing it?

Reply:
Also can the Source be the output of the previous task?

Reply:
> is this only intended to support git commits or gcs buckets? maybe a generic interface with git and gcs types implementing it?

Definitely we want to be able to support other sources than just git! It's not clear how exactly to design this though since some tasks might have logic that expects a git commit hash (e.g. naming an image)

I'll add an issue to our issue tracker shortly so we can dive into this more!

> Also can the Source be the output of the previous task?

Oh neat! In the current design we didn't allow for that but do you think that's something folks would use?

Reply:
Can you let us know your thoughts about the other types of `Sources` and how to support a generic interface?

Yes, I think its useful to for example be able to `build an image` in a step and `run tests on that image just built before uploading it` in the following step. This would be need the ability to an input of a task to be a output of a previous task

Reply:
This makes it harder to make sure the `testImage` step is actually testing the image that was meant to be created in the `buildPush` step or picked up an older image because the push failed.

Reply:
> This makes it harder to make sure the testImage step is actually testing the image that was meant to be created in the buildPush step or picked up an older image because the push failed.

That's a great point. What do you think about being sure that the image is correct by making the digest available? @tejal29 has been suggesting we add an `image` type that could be an `output` and an in `input` - so a Task that builds an image could expose an `image`, including the digest, and the test Task could be given that image as an input via the Pipeline. (We also want to make sure we make use of the work already done to [design build artifacts](https://docs.google.com/document/d/1FUn9LZXUAJQb8atoTeRxmsdjxpstOyegHF7ieJEblNs/edit#) _note doc shared with knative-dev@googlegroups.com_).

One downside though is the extra overhead to upload the image to a registry in the build Task and then download it in the test Task :(

Reply:
The issue of knowing that the sources/inputs you're operating on throughout the pipeline stay the same came up at the build-crd working group meeting also, so maybe we should tackle the issue more generally as well :thinking:

Reply:
Yeah, I agree the generic type of input/output might be the way to think about this, an `image` type can be one implementation, and `git` another. 

We would need to indicate to a `taskRun` somehow that its input is the output of another taskRun

> One downside though is the extra overhead to upload the image to a registry in the build  Task and then download it in the test Task :(

Wouldn't we need to do that anyway for the test task to run against the new image?

I think your suggestion of having another issue to cover the design of Source input/output might be the best option to move this forward

Reply:
Created https://github.com/knative/build-pipeline/issues/11 to discuss particularly the issue of being sure you're using the sources/inputs you expect to be using.

Reply:
Also created #13 to explore sources in more detail!